rfcs/45-stageless.md

```diff
@@ -0,0 +1,980 @@
+A system or system set will only run if all of its conditions return `true`.
+If one of its conditions returns `false`, the system (or members of the system set) will be skipped.
+
+To be clear, systems can have multiple conditions, and those conditions are not shared with others.
+Each condition instance is unique and will be evaluated *at most once* per run of a schedule.
+Conditions are evaluated right before their guarded system (or the first system in their guarded system set that's able to run) would be run, so their results are guaranteed to be up-to-date.
+The data read by conditions will not change before the guarded system starts.
+
+```rust
+// This is just an ordinary system: timers need to be ticked!
+fn tick_construction_timer(timer: ResMut<ConstructionTimer>, time: Res<Time>){
+    timer.tick(time.delta());
+}
+
+// This function can serve as a condition because it does not have mutable access and it returns `bool`.
+fn construction_timer_not_finished(timer: Res<ConstructionTimer>) -> bool {
+    !timer.finished()
+}
+
+// As long as they're immutable, you can use queries, event readers and resources with arbitrarily complex internal logic.
+fn too_many_enemies(
+    population_query: Query<(), With<Enemy>>,
+    population_limit: Res<PopulationLimit>,
+) -> bool {
+    population_query.iter().count() > population_limit
+}
+
+fn main() {
+    App::new()
+        .add_plugins(DefaultPlugins)
+        // You can add functions with read-only system parameters as conditions.
+        .add_set(GameSet::Construction.run_if(construction_timer_not_finished))
+        .add_systems(
+            systems![tick_construction_timer, update_construction_progress]
+                .in_set(GameSet::Construction))
+        .add_system(mitigate_meltdown.run_if(too_many_enemies))
+        // You can use closures for simple one-off conditions.
+        .add_system(spawn_more_enemies.run_if(|difficulty: Res<Difficulty>| difficulty >= 9000))
+        // `resource_exists` and `resource_equals` are helper functions that produce
+        // new closures that can be used as conditions.
+        .add_system(gravity.run_if(resource_exists(Gravity)))
+        // The systems in this set won't run if `Paused` is `true`.
+        .add_set(GameSet::Physics.run_if(resource_equals(Paused(false))))
+        .run();
+}
+```
+
+### Exclusive systems
+
+As much as everyone loves to see systems running in parallel, sometimes a little `&mut World` is necessary.
+Previous versions of Bevy treated these "exclusive" systems as a separate concept, with their own set of types and traits.
+They couldn't have `Local` params and could only be inserted at specific points of a stage.
+
+That is no longer the case.
+Exclusive systems are now just regular systems.
+They can have `Local` params and be scheduled wherever you want.
+
+Command application is now signaled through an exclusive system called `apply_system_buffers`.

```

Comment:
Personally, I think a name like `apply_commands` or something similar would be more intuitive. Commands should be known to every user as a concept, naming it after something more abstract like buffers could cause confusion.

Does this system only handle commands or does it do other things too?

Reply:
It's called `apply_system_buffers` because some way or another, it'll call the `apply_buffers` method on your systems (which are `Box<dyn System>` trait objects). Pretty literal.

There is no method on the `System` trait that only applies commands. `Commands` just happens to be the only built-in `SystemParam` that has buffers, so it works out.